Implement Charged Candidate Maker (PFA1b)

ruse-traveler opened the pull request:
### Briefly, what does this PR introduce?

This PR implements the `ChargedCandidateMaker` algorithm. This algorithm ingests track-cluster matches and converts them to charged particle candidates, i.e. `edm4eic::ReconstructedParticles` with only select relations filled.

### What kind of change does this PR introduce?
- [ ] Bug fix (issue #__)
- [x] New feature (issue #2128  )
- [ ] Documentation update
- [ ] Other: __

### Please check if this PR fulfills the following:
- [ ] Tests for the changes have been added
- [ ] Documentation has been added / updated
- [x] Changes have been communicated to collaborators

### Does this PR introduce breaking changes? What changes might users need to make to their code?

No.

### Does this PR change default behavior?

No.

ruse-traveler committed:
Add skeletons of algorithm and factory

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

ruse-traveler committed:
Merge branch 'main' into add-track-cluster-subtraction-pfa-one

ruse-traveler committed:
Propagate upstream changes from development fork

ruse-traveler committed:
Merge branch 'add-track-cluster-subtraction-pfa-one' of github.com:eic/EICrecon into add-track-cluster-subtraction-pfa-one

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

ruse-traveler committed:
Load particle plugin

ruse-traveler committed:
Use track PDGs for masses

ruse-traveler committed:
Fix merge conflicts

ruse-traveler committed:
Allow for checking zero against resolutions

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

ruse-traveler committed:
Write out all relevant collections

ruse-traveler committed:
Fix merge conflicts

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

ruse-traveler committed:
Merge branch 'main' into add-track-cluster-subtraction-pfa-one

ruse-traveler committed:
Add particle category to global reconstruction

ruse-traveler committed:
Merge branch 'add-track-cluster-subtraction-pfa-one' of github.com:eic/EICrecon into add-track-cluster-subtraction-pfa-one

ruse-traveler committed:
Merge main

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

wdconinc committed:
Implement Track-Cluster Subtraction (PFA1) (fix: iwyu) (#2051)

This PR applies the include-what-you-use fixes as suggested by
https://github.com/eic/EICrecon/actions/runs/17247257038.
Please merge this PR into the branch
`add-track-cluster-subtraction-pfa-one`
to resolve failures in PR #1627.

Auto-generated by [create-pull-request][1]

[1]: https://github.com/peter-evans/create-pull-request

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

ruse-traveler committed:
Remove PFTools

ruse-traveler committed:
Make sure cmake finds plugin

ruse-traveler committed:
Use track-cluster matches until merge/splitter is updated

ruse-traveler committed:
Make consistent with reco CMakeLists

ruse-traveler committed:
Do light reformatting

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

ruse-traveler committed:
Fix merge conflicts

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

ruse-traveler committed:
Merge branch 'main' into add-track-cluster-subtraction-pfa-one

ruse-traveler committed:
Remove unused ChangeRun, make Process consistent with base class

ruse-traveler committed:
Add outputs to defaults

ruse-traveler committed:
Clean up, add header guards

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

wdconinc committed:
Implement Track-Cluster Subtraction (PFA1) (fix: iwyu) (#2087)

This PR applies the include-what-you-use fixes as suggested by
https://github.com/eic/EICrecon/actions/runs/17839250702.
Please merge this PR into the branch
`add-track-cluster-subtraction-pfa-one`
to resolve failures in PR #1627.

Auto-generated by [create-pull-request][1]

[1]: https://github.com/peter-evans/create-pull-request

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

ruse-traveler committed:
Merge branch 'main' into add-track-cluster-subtraction-pfa-one

ruse-traveler committed:
Merge branch 'main' into add-track-cluster-subtraction-pfa-one

ruse-traveler committed:
Add FHCal insert

ruse-traveler committed:
Make sure FHCal insert collections are in default output

ruse-traveler committed:
Rename factories after primary output

ruse-traveler committed:
Remove WIP notes

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

ruse-traveler committed:
Add algorithm, factory, and generators

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

ruse-traveler committed:
Remove edm4eic 8.X guards

ruse-traveler committed:
Merge main

ruse-traveler committed:
Merge PFA1(a) branch

ruse-traveler committed:
Remove edm4eic 8.X guards

ruse-traveler committed:
Remove unnecessary comments

ruse-traveler committed:
Add missing factory

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

ruse-traveler committed:
Fix typo in docstring

ruse-traveler committed:
Fix typo in docstring

ruse-traveler committed:
Merge branch 'add-track-cluster-subtraction-pfa-one' into add-charged-candidate-maker-pfa-one

wdconinc committed:
Implement Track-Cluster Subtraction (PFA1a) (fix: iwyu) (#2133)

This PR applies the include-what-you-use fixes as suggested by
https://github.com/eic/EICrecon/actions/runs/18477471601.
Please merge this PR into the branch
`add-track-cluster-subtraction-pfa-one`
to resolve failures in PR #1627.

Auto-generated by [create-pull-request][1]

[1]: https://github.com/peter-evans/create-pull-request

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

wdconinc committed:
Implement Charged Candidate Maker (PFA1b) (fix: iwyu) (#2134)

This PR applies the include-what-you-use fixes as suggested by
https://github.com/eic/EICrecon/actions/runs/18477497612.
Please merge this PR into the branch
`add-charged-candidate-maker-pfa-one`
to resolve failures in PR #2124.

Auto-generated by [create-pull-request][1]

[1]: https://github.com/peter-evans/create-pull-request

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

veprbl commented on src/algorithms/particle/ChargedCandidateMaker_factory.h:
Duplicate file? All factories go to factories/

veprbl reviewed (commented):
Not sure about the particle/ namespace instead of reco/, but we can have it.

ruse-traveler commented on src/global/particle/particle.cc:
```diff
@@ -90,5 +99,56 @@ void InitPlugin(JApplication* app) {
+
+  app->Add(new JOmniFactoryGeneratorT<ChargedCandidateMaker_factory>(
+      "EndcapPInsertChargedCandidateParticlesAlpha", {"EndcapPInsertTrackExpectedClusterMatches"},
+      {"EndcapPInsertChargedCandidateParticlesAlpha"}, {}, app));
```
Sounds good!

ruse-traveler committed:
Remove misplaced factory

ruse-traveler committed:
Temporarily add charged candidates to default output

ruse-traveler committed:
Merge branch 'main' into add-track-cluster-subtraction-pfa-one

ruse-traveler committed:
Merge branch 'main' into add-charged-candidate-maker-pfa-one

ruse-traveler committed:
Merge PFA1a branch into PFA1b

ruse-traveler commented:
> @ruse-traveler I'm noticing this is targeting add-charged-candidate-maker-pfa-one branch instead of main. Is that intentional?

Ah! That was intentional: I wasn't sure what order we would want to review these in, so I pointed it at the PFA1(a) branch in the interim. It's extremely straight-forward to swap it around and point it at main, though!

ruse-traveler committed:
Use track-cluster matches as PFA1b input

ruse-traveler committed:
Remove PFA1a code

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

veprbl commented on src/factories/particle/ChargedCandidateMaker_factory.h:
```diff
@@ -0,0 +1,42 @@
+  ///! Configures algorithm
+  void Configure() {
+    m_algo = std::make_unique<AlgoT>(GetPrefix());
+    m_algo->init();
```
```suggestion
    m_algo->level(static_cast<algorithms::LogLevel>(logger()->level()));
    m_algo->applyConfig(config());
    m_algo->init();
```
(config is empty, but it's fine to apply for consistency)

ruse-traveler commented on src/algorithms/particle/ChargedCandidateMaker.h:
```diff
@@ -0,0 +1,71 @@
+ *  tracks, subtracts the sum of all tracks pointing to the cluster,
+ *  and outputs the remnant clusters, expected clusters, and their matched
+ *  tracks.
+ */
```
Oop! Good catch!

ruse-traveler committed:
Fix typos in docstrings

ruse-traveler committed:
Merge branch 'main' into add-charged-candidate-maker-pfa-one

ruse-traveler committed:
Add log level/config application to factory init for consistency

ruse-traveler commented on src/factories/particle/ChargedCandidateMaker_factory.h:
```diff
@@ -0,0 +1,44 @@
+
+  // output collection
+  PodioOutput<edm4eic::ReconstructedParticle> m_charged_candidate_output{this};
+
```
```suggestion

  // services
  Service<AlgorithmsInit_service> m_algoInitSvc{this};
```

ruse-traveler committed:
Don't forget init service

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

ruse-traveler committed:
Revert "Don't forget init service"

This reverts commit bb98137270d9652d597bf8d705d8da0841c33764.

veprbl committed:
Merge branch 'main' into add-charged-candidate-maker-pfa-one

veprbl commented on src/global/particle_flow/particle.cc:
```diff
@@ -0,0 +1,79 @@
+  app->Add(
+      new JOmniFactoryGeneratorT<CollectionCollector_factory<edm4eic::TrackClusterMatch, false>>(
+          "EndcapPInsertTrackClusterMatches",
+          {"EcalEndcapPTrackClusterMatches", "HcalEndcapPInsertTrackClusterMatches"},
```
If we use Ecal for both LFHCAL and insert, that's double counting the candidates?

ruse-traveler replied:
Ah! Good catch! I wanted to make sure and include the insert since there is overlap in coverage between it and the FEMC, but I'm not sure why I didn't just include the insert in the `EndcapPTrackClusterMatches`.

My suspicion is that it's pretty rare for a track to be simultaneously matched to an FEMC, LFHCAL, and insert cluster. But if we include all 3 in the same map, then that will at least make sure we're not double counting the tracks in these cases...

veprbl commented on src/algorithms/particle/ChargedCandidateMaker.cc:
```diff
@@ -0,0 +1,50 @@
+  auto [out_particle]   = output;
+
+  // exit if no matches in collection
+  if (in_match->size() == 0) {
```
```suggestion
  if (in_match->empty()) {
```
clang-tidy is not picking this up because it's not an STL collection.

veprbl commented on src/algorithms/particle/ChargedCandidateMaker.h:
```diff
@@ -0,0 +1,70 @@
+  //! Comparator struct for tracks
+  // ------------------------------------------------------------------------
+  /*! Organizes tracks by their ObjectIDs in decreasing collection
+   *  ID first, and by decreasing index second.
```
```suggestion
  /*! Organizes tracks by their ObjectIDs in increasing collection
   *  ID first, and by increasing index second.
```

veprbl committed:
add a unit test, add needed NoConfig

epic-capybara committed:
Implement Charged Candidate Maker (PFA1b) (fix: iwyu) (#2649)

This PR applies the include-what-you-use fixes as suggested by
https://github.com/eic/EICrecon/actions/runs/25325316004.
Please merge this PR into the branch
`add-charged-candidate-maker-pfa-one`
to resolve failures in PR #2124.

Auto-generated by [create-pull-request][1]

[1]: https://github.com/peter-evans/create-pull-request

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

ruse-traveler committed:
Fix typo in comparator struct

Co-authored-by: Dmitry Kalinkin <dmitry.kalinkin@gmail.com>

ruse-traveler committed:
Prefer empty() over checking size

Co-authored-by: Dmitry Kalinkin <dmitry.kalinkin@gmail.com>

ruse-traveler committed:
Merge branch 'add-charged-candidate-maker-pfa-one' of github.com:eic/EICrecon into add-charged-candidate-maker-pfa-one

ruse-traveler committed:
Rename plugin to match other stages

ruse-traveler committed:
Make CMakeLists consistent with PFA1a

ruse-traveler committed:
Process FEMC, LFHCAL, and insert in one algorithm call

ruse-traveler committed:
Use CompareObjectID

ruse-traveler committed:
Clean up: update copyrights, remove unnecessary comments

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

ruse-traveler commented:
Just pushed a few more changes to update the plugin name, process forward calos in one call, and use `CompareObjectID` @veprbl !

Silly question: I'm seeing some odd changes wrt main for changes I know have already been merged (e.g. adding a note to the "high priority" item on the PR template)... Should I do a rebase?

veprbl committed:
Merge branch 'main' into add-charged-candidate-maker-pfa-one

veprbl committed:
Delete src/global/particle_flow/particle.cc

epic-capybara committed:
Implement Charged Candidate Maker (PFA1b) (fix: iwyu) (#2657)

This PR applies the include-what-you-use fixes as suggested by
https://github.com/eic/EICrecon/actions/runs/25382062240.
Please merge this PR into the branch
`add-charged-candidate-maker-pfa-one`
to resolve failures in PR #2124.

Auto-generated by [create-pull-request][1]

[1]: https://github.com/peter-evans/create-pull-request

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

veprbl reviewed (approved):
LGTM
